distsql: specify stream schema in input synchronizer

RaduBerinde opened the pull request:
Currently the distsql flow specs do not include the types of the data going into
each processor. This makes things like adding filters more difficult
(initializing a filtering expression requires the types). Some of the processors
(aggregation, hashing) require this information so it is part of their specs.

This changes adds the schema information to InputSyncSpec and to the `RowSource`
interface so that any processor can get the input schema based on their sources.

This also introduces the possibility of cross-checking that the data moving
through a stream (internal or external) matches the types envisioned by the
planner; however we are not adding this at this time since it may be expensive.

-----------
Important to review are the `.proto`, `base.go` and `distsql_physical_planning.go` changes. The rest of the changes are kind of mindless.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12932)
<!-- Reviewable:end -->

RaduBerinde committed:
distsql: specify stream schema in input synchronizer

Currently the distsql flow specs do not include the types of the data going into
each processor. This makes things like adding filters more difficult
(initializing a filtering expression requires the types). Some of the processors
(aggregation, hashing) require this information so it is part of their specs.

This changes adds the schema information to InputSyncSpec and to the `RowSource`
interface so that any processor can get the input schema based on their sources.

This also introduces the possibility of cross-checking that the data moving
through a stream (internal or external) matches the types envisioned by the
planner; however we are not adding this at this time since it may be expensive.

RaduBerinde committed:
distsql: rename Expr to Aggregation in AggregatedSpec

I think this name was left over form an early iteration of the change where it
actually was an expression. Changing to a less confusing name.

RaduBerinde committed:
distsql: evaluator: move expr init out of the main loop

Now that we have the input types in the `RowSource` the evaluator doesn't need
to wait for the first row before initializing the expressions.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 964 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahTVOQi3URKcZk2lM1:-KahTVOQi3URKcZk2lM2:bsjab9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsql_physical_planner.go#L964)):*
> ```Go
> 		types := make([]sqlbase.ColumnType, len(finalAggSpec.Aggregations))
> 		for i, agg := range finalAggSpec.Aggregations {
> 			_, types[i], err = distsqlrun.GetAggregateInfo(agg.Func, inputTypes[agg.ColIdx])
> ```

should we move this `GetAggregateInfo()` out of `distsqlrun`? Into some sort of `base`.

---

*[pkg/sql/distsqlrun/base.go, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahMcuOBf-8AK4rxH2C:-KahMcuOBf-8AK4rxH2D:bi7njf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsqlrun/base.go#L282)):*
> ```Go
> }
> 
> // Types is part of the RowSource interface.
> ```

hmmm but what if this is called when there's no rows in the buffer?
Should we pass the schema to the ctor? In fact, I think you've had to add that `EmptyRowSrc` to a test for this reason...

---

*[pkg/sql/distsqlrun/data.proto, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahLw3jXUUiBQgkLjGv:-KahLw3k1n_67UMpZjgx:bf5cqla) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsqlrun/data.proto#L106)):*
> ```protobuf
>   // Schema for the streams entering this synchronizer.
>   repeated sqlbase.ColumnType column_types = 4 [(gogoproto.nullable) = false];
> ```

nit: extra line

---

*[pkg/sql/distsqlrun/evaluator.go, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahUGUA7v5atDMT59r5:-KahUGUA7v5atDMT59r6:b-79dqof) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a47c8190a596378d7eae3d2448f842ffbd0c9a/pkg/sql/distsqlrun/evaluator.go#L57)):*
> ```Go
> 
> 	types := input.Types()
> 	for i, expr := range ev.specExprs {
> ```

perhaps a comment about what this block does

---

*[pkg/sql/distsqlrun/evaluator.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahUO-YD665n10dPUtj:-KahUO-YD665n10dPUtk:b-29hkuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffc0c336351e06b68e7982b5ac6008ba75aa0a66/pkg/sql/distsqlrun/evaluator.go#L91)):*
> ```Go
> 				err := ev.exprs[i].init(expr, types, ev.flowCtx.evalCtx)
> 				if err != nil {
> 					ev.output.Close(err)
> ```

nice that we got rid of this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-:-KahUR419LglKXXubk-X:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 964 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahTVOQi3URKcZk2lM1:-KahbOOGn-lUZom8ffyZ:b-arw4mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsql_physical_planner.go#L964)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

should we move this `GetAggregateInfo()` out of `distsqlrun`? Into some sort of `base`.
</blockquote></details>

It's associated with `AggregatorSpec_Func` so it belongs in the same package. If we ever move the specs to a base package, indeed it would belong there.

---

*[pkg/sql/distsqlrun/base.go, line 282 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahMcuOBf-8AK4rxH2C:-Kahp14ZuzrvCGdlei1A:bnc8a06) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsqlrun/base.go#L282)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

hmmm but what if this is called when there's no rows in the buffer?
Should we pass the schema to the ctor? In fact, I think you've had to add that `EmptyRowSrc` to a test for this reason...
</blockquote></details>

Added constructor that takes types (optional if there are rows we can use to deduce them).

---

*[pkg/sql/distsqlrun/data.proto, line 106 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahLw3jXUUiBQgkLjGv:-KahpgHsABzM0dPsR8bn:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4b3ccecf24628d2714fffb443a2909a9f314a7f/pkg/sql/distsqlrun/data.proto#L106)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: extra line
</blockquote></details>

Fixed.

---

*[pkg/sql/distsqlrun/evaluator.go, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-KahUGUA7v5atDMT59r5:-Kahpoe-mS5mOmj_rzW8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a47c8190a596378d7eae3d2448f842ffbd0c9a/pkg/sql/distsqlrun/evaluator.go#L57)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

perhaps a comment about what this block does
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12932#-:-KahpqAgaTx7rDgPyTeP:b-814g8k)*
<!-- Sent from Reviewable.io -->
